File: verify-node/handler.js
'use strict'
const http = require('http');
const crypto = require('crypto');
const hash = crypto.createHash('sha256');

module.exports = (body, callback) => {
    var data = '';

    http.get('http://gateway:8080/certificates/callback', (res) => {
        res.setEncoding('utf8');
        res.on('data', function(chunk) {
            data += chunk;
        });

        res.on('end', () => {
            let json = JSON.parse(data);
            verify(json.pem);
        });
    }).on('error', (e) => {
        console.warn(`Got error: ${e.message}`);
        callback(undefined, {status: 'error', error: e.message});
    });

    function verify(publicKey) {
        let auth = process.env.Http_Authorization;
        if (auth === undefined) {
            callback(undefined, {status: 403, message: 'missing authorization header'});
            return
        }

        let signingString = getSigningString(body);
        let signature = getSignature(auth);
        console.warn(sprintf('signing string:\n%s', signingString));
        console.warn('public key:\n' + publicKey);
        console.warn('authorization: %s\n', auth);
        console.warn('signature: %s\n', signature);

        var verify = crypto.createVerify('RSA-SHA256');
        verify.update(signingString);

        var verification = verify.verify(publicKey, signature, 'base64');

        if (verification) {
            console.warn("Verification OK");
            callback(undefined, {status: '200', verification: verification});

        } else {
            console.warn("Verification failed");
            callback(undefined, {status: '401', verification: verification});
        }
    }

    function getSigningString(body) {

        hash.update(body);
        let digest = hash.digest('base64');

        let res = '';
        res += sprintf('%s\n', getRequestTarget());
        res += sprintf('host: %s\n', process.env.Http_X_Forwarded_Host);
        res += sprintf('date: %s\n', process.env.Http_Date);
        res += sprintf('content-type: %s\n', process.env.Http_Content_Type);
        res += sprintf('digest: SHA-256=%s\n', digest);
        res += sprintf('content-length: %s', body.length);

        return res
    }

    function getSignature(authHeader) {
        return authHeader.split(',')[3].split('="')[1] + '='
    }

    function getRequestTarget() {
        let path = process.env.Http_Path;
        if (path === '/') {
            path = '';
        }

        let name = process.env.Http_Host.split(':')[0];

        let requestTarget = '(request-target): ' + process.env.Http_Method.toLowerCase() + ' /function/' + name + path;
        let query = process.env.Http_Query;
        if (query !== undefined) {
            requestTarget = requestTarget + '?' + query
        }

        return requestTarget
    }

    function sprintf(str) {
        var args = [].slice.call(arguments, 1),
            i = 0;

        return str.replace(/%s/g, () => args[i++]);
    }
};

File: verify-python/handler.py
import base64
import os
import hashlib
import logging
import requests

def handle(req):

    """handle a request to the function
    Args:
        req (str): request body
    """

    body_digest = get_digest(req)
    signing_string = get_signing_string(req, body_digest)
    signature = get_signature()
    public_key = get_public_key()

    logging.warn('\n' + signing_string)
    logging.warn('\n' + public_key)

    if verify(signing_string, public_key, signature):
        logging.warn('{status: "200", verification: true}')
        return '{status: "200", verification: true}'
    else:
        logging.warn('{status: "200", verification: false}')
        return '{status: "200", verification: false}'


def verify(signing_string, public_key_pem, signature):
    from Crypto.PublicKey import RSA
    from Crypto.Signature import PKCS1_v1_5
    from Crypto.Hash import SHA256
    from base64 import b64decode

    public_key = RSA.importKey(public_key_pem)
    signer = PKCS1_v1_5.new(public_key)
    digest = SHA256.new()
    digest.update(signing_string)

    if signer.verify(digest, b64decode(signature)):
        return True
    return False


def get_signature():
    auth_header = os.environ['Http_Authorization']
    signature = auth_header.split(',')[3].split('="')[1] + '='

    return signature


def get_digest(body):
    sha_hash = hashlib.sha256()
    sha_hash.update(body)
    digest = base64.b64encode(sha_hash.digest()).decode()

    return digest


def get_signing_string(body, digest):

    res = ''
    res += '{}\n'.format(get_request_target())
    res += 'host: {}\n'.format(os.environ['Http_X_Forwarded_Host'])
    res += 'date: {}\n'.format(os.environ['Http_Date'])
    res += 'content-type: {}\n'.format(os.environ['Http_Content_Type'])
    res += 'digest: SHA-256={}\n'.format(digest)
    res += 'content-length: {}'.format(len(body))
    
    return res


def get_request_target():

    path = os.environ['Http_Path']
    if path == '/':
        path = ''

    name = os.environ['Http_Host'].split(':')[0]

    request_target = '(request-target): ' + os.environ['Http_Method'].lower() + ' /function/' + name + path

    if 'Http_Query' in os.environ:
        query = os.environ['Http_Query']
        request_target = request_target + '?' + query

    return request_target


def get_public_key():

    req = requests.get('http://gateway:8080/certificates/callback')

    return req.json()['pem']
